utils/writeable/src/utf.rs

```diff
@@ -0,0 +1,89 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{impl_display_with_writeable, LengthHint, Writeable};
+
+use core::fmt;
+
+/// Implements [`Writeable`] for [`&[u8]`] according to the [WHATWG Encoding Standard](
+/// https://encoding.spec.whatwg.org/#utf-8-decoder).
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)] // newtype
+pub struct PotentiallyInvalidUtf8<'a>(pub &'a [u8]);
+
+impl Writeable for PotentiallyInvalidUtf8<'_> {
+    fn write_to<W: core::fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        let mut remaining = self.0;
+        while !remaining.is_empty() {
+            match core::str::from_utf8(remaining) {
+                Ok(str) => {
+                    return sink.write_str(str);
+                }
+                Err(e) => {
+                    let (str, rest) = remaining.split_at(e.valid_up_to());
+                    sink.write_str(unsafe { core::str::from_utf8_unchecked(str) })?;
```

Comment:
Nit: Include safety comment so future readers don't need to go to the `core::str::from_utf8` docs page to read about the invariants.

Thought: It would be nice if `core::str::Utf8Error` had a safe function to return the split for you.